File: 배열/leetcode605(Easy)-Can Place Flowers.js
/**
 * @param {number[]} flowerbed
 * @param {number} n
 * @return {boolean}
 */
var canPlaceFlowers = function (flowerbed, n) {
  if (n === 0) {
    return true
  }

  const extended = [0, ...flowerbed, 0] // 굿 아이디어

  for (let i = 1; i < flowerbed.length + 1; i += 1) {
    if (extended[i] === 1) {
      continue
    }
    if (extended[i - 1] === 0 && extended[i + 1] === 0) {
      extended[i] = 1
      n -= 1
      if (n === 0) {
        return true
      }
    }
  }

  return false
}

File: 배열/leetcode605(Easy)-Can Place Flowers.py
"""
<나의 풀이>
접근 : 그리디 + 배열 문제
- 시간복잡도 : O(N)
- 공간복잡도 : O(1)
"""
class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        def can_plant(index):
            if flowerbed[index] == 1:
                return False

            if index == 0:
                if index + 1 >= len(flowerbed):
                    return True
                return flowerbed[index + 1] == 0
            
            if index == len(flowerbed) - 1:
                if index - 1 < 0:
                    return True
                return flowerbed[index - 1] == 0
            
            return flowerbed[index - 1] == 0 and flowerbed[index + 1] == 0

        if n == 0:
            return True

        for i in range(len(flowerbed)):
            if can_plant(i):
                flowerbed[i] = 1
                n -= 1
                if n == 0:
                    return True

        return False
    

"""
굿 아이디어 : 범위를 3군데로 지정해주는 것이 귀찮을 수 있는데, 이렇게 맨 앞, 맨 뒤에 빈 값을 넣게되면 매우 심플하게 풀이가 가능.
"""
class Solution:
    def canPlant(self, index, arr):
        if arr[index] == 1:
            return False
        return arr[index - 1] == 0 and arr[index + 1] == 0

    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        if n == 0:
            return True
        
        extended = [0] + flowerbed + [0]
        for i in range(1, len(flowerbed) + 1):
            if self.canPlant(i, extended):
                extended[i] = 1
                n -= 1
                if n == 0:
                    return True

        return False